test/unit/org/apache/cassandra/io/filesystem/ListenableFileSystem.java

```diff
@@ -0,0 +1,541 @@
+    }
+
+    public interface OnTransferTo extends Listener
+    {
+        default void preTransferTo(Path path, FileChannel channel, long position, long count, WritableByteChannel target) throws IOException
+        {
+        }
+
+        void postTransferTo(Path path, FileChannel channel, long position, long count, WritableByteChannel target, long transfered) throws IOException;
+    }
+
+    public interface OnTransferFrom extends Listener
+    {
+        default void preTransferFrom(Path path, FileChannel channel, ReadableByteChannel src, long position, long count) throws IOException
+        {
+        }
+
+        void postTransferFrom(Path path, FileChannel channel, ReadableByteChannel src, long position, long count, long transfered) throws IOException;
+    }
+
+    public interface OnWrite extends Listener
+    {
+        void preWrite(Path path, FileChannel channel, long position, ByteBuffer src) throws IOException;
+
+        void postWrite(Path path, FileChannel channel, long position, ByteBuffer src, int wrote) throws IOException;
+    }
+
+    public interface OnChannelMeta extends Listener
+    {
+        default void prePosition(Path path, FileChannel channel, long position, long newPosition) throws IOException
+        {
+        }
+
+        default void postPosition(Path path, FileChannel channel, long position, long newPosition) throws IOException
+        {
+        }
+
+        default void preTruncate(Path path, FileChannel channel, long size, long targetSize) throws IOException
+        {
+        }
+
+        default void postTruncate(Path path, FileChannel channel, long size, long targetSize, long newSize) throws IOException
+        {
+        }
+
+        default void preForce(Path path, FileChannel channel, boolean metaData) throws IOException
+        {
+        }
+
+        default void postForce(Path path, FileChannel channel, boolean metaData) throws IOException
+        {
+        }
+    }
+
+    public interface Unsubscribable extends AutoCloseable
+    {
+        @Override
+        void close();
+    }
+
+    private final List<OnOpen> onOpen = new CopyOnWriteArrayList<>();
+    private final List<OnTransferTo> onTransferTo = new CopyOnWriteArrayList<>();
+    private final List<OnRead> onRead = new CopyOnWriteArrayList<>();
+    private final List<OnWrite> onWrite = new CopyOnWriteArrayList<>();
+    private final List<OnTransferFrom> onTransferFrom = new CopyOnWriteArrayList<>();
+    private final List<OnChannelMeta> onChannelMeta = new CopyOnWriteArrayList<>();
+    private final List<List<? extends Listener>> lists = Arrays.asList(onOpen, onRead, onTransferTo, onWrite, onTransferFrom, onChannelMeta);
+    private final ListenableFileSystemProvider provider;
+
+    public ListenableFileSystem(FileSystem delegate)
+    {
+        super(delegate);
+        this.provider = new ListenableFileSystemProvider(super.provider());
+    }
+
+    public Unsubscribable listen(Listener listener)
```

Comment:
feedback *strongly* welcomed on this API as I actually hate it... but not sure what would be better...  want to expose every call to file system so tests can do w/e they want, but current one is rather ugly to work with...